xdsl/dialects/onnx.py

```diff
@@ -671,6 +672,145 @@ def verify_(self) -> None:
+    to 0 along start and end of each spatial axis.
+
+    - storage_order int (default is '0') : The storage order of the tensor. 0 is row major, and 1 is column major.
+    This attribute is used only to convert an n-tuple index value into a single integer value for producing the
+    second output.
+
+    - strides list of ints: Stride along each spatial axis. If not present, the stride defaults to 1 along each spatial axis
+
+    """
+
+    name = "onnx.MaxPoolSingleOut"
+
+    T = Annotated[AnyFloat | IntegerType, ConstraintVar("T")]
+    data = operand_def(TensorType[T] | MemRefType[T])
+    output = result_def(TensorType[T] | MemRefType[T])
+
+    auto_pad = attr_def(StringAttr)
+    ceil_mode = attr_def(AnyIntegerAttr)
+    dilations = attr_def(ArrayAttr[AnyIntegerAttr])
+    kernel_shape = attr_def(ArrayAttr[AnyIntegerAttr])
+    pads = attr_def(ArrayAttr[AnyIntegerAttr])
+    storage_order = attr_def(AnyIntegerAttr)
+    strides = attr_def(ArrayAttr[AnyIntegerAttr])
+
+    assembly_format = (
+        "`(` $data`)` attr-dict `:` `(` type($data) `)` `->` type($output)"
+    )
+
+    def __init__(
+        self,
+        data: SSAValue,
+        auto_pad: Attribute,
+        ceil_mode: Attribute,
+        dilations: Attribute,
+        kernel_shape: Attribute,
+        pads: Attribute,
+        storage_order: Attribute,
+        strides: Attribute,
+    ):
+        super().__init__(
+            attributes={
+                "auto_pad": auto_pad,
+                "ceil_mode": ceil_mode,
+                "dilations": dilations,
+                "kernel_shape": kernel_shape,
+                "pads": pads,
+                "storage_order": storage_order,
+                "strides": strides,
+            },
+            operands=[data],
+            result_types=[data.type],
+        )
+
+    def verify_(self) -> None:
+        if not isinstance(
+            data_type := self.data.type, TensorType | MemRefType
+        ) or not isinstance(output_type := self.output.type, TensorType | MemRefType):
+            assert False, (
+                "onnx elementwise operation operands (data) and result (output) must be of type TensorType or "
+                "MemRefTyoe "
+            )
+
+        data_type = cast(TensorType[Attribute], data_type)
+        output_type = cast(TensorType[Attribute], output_type)
+
+        # auto_pad
+        auto_pad_strings = ["NOTSET", "SAME_UPPER", "SAME_LOWER", "VALID"]
+        if self.auto_pad.data not in auto_pad_strings:
+            raise VerifyException(
+                f"Invalid auto_pad string. Must be one of {auto_pad_strings}"
+            )
+
+        # ceil mode
+        if self.ceil_mode.value.data < 0 or self.ceil_mode.value.data > 1:
+            raise VerifyException("ceil value must be either zero or one")
+
+        # kernel shape
+        if len(data_type.get_shape()) - 2 != len(self.kernel_shape):
+            raise VerifyException("input data and kernel shape rank mismatch ")
```

Comment:
Can you please an f-string here (and below where appropriate) giving more information about the lengths provided? You can use the `:=` to store the `len` result within the if condition and use it just within the scope of its body.